modules/store/src/reducer_creator.ts

```diff
@@ -1,17 +1,28 @@
 import { ActionCreator, ActionReducer, ActionType, Action } from './models';
 
+// Goes over the array of ActionCreators, pulls the action type out of each one
+// and returns the array of these action types.
+type ExtractActionTypes<Creators extends readonly ActionCreator[]> = {
+  [Key in keyof Creators]: Creators[Key] extends ActionCreator<infer T>
+    ? T
+    : never;
+};
+
 /**
  * Return type of the `on` fn.
  * Contains the action reducer coupled to one or more action types.
  */
-export interface ReducerTypes<S> {
-  reducer: ActionReducer<S>;
-  types: string[];
+export interface ReducerTypes<
+  State,
+  Creators extends readonly ActionCreator[]
+> {
+  reducer: OnReducer<State, Creators>;
+  types: ExtractActionTypes<Creators>;
 }
 
 // Specialized Reducer that is aware of the Action type it needs to handle
-export interface OnReducer<S, C extends ActionCreator[]> {
-  (state: S, action: ActionType<C[number]>): S;
+export interface OnReducer<S, C extends readonly ActionCreator[]> {
+  (state: S, action: ActionType<C[number]>): { [P in keyof S]: S[P] };
```

Comment:
Since we're renamed all `S` types to `State`, should we also do it here?

```suggestion
export interface OnReducer<State, Creators extends readonly ActionCreator[]> {
  (state: State, action: ActionType<Creators[number]>): { [P in keyof State]: State[P] };
```